proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,531 @@
+
+* intra-process: messages are sent from a publication to subscription via in-process memory
+* inter-process: messages are serialized and sent via the underlying middleware.
+
+This design document presents a new implementation for the intra-process communication.
+
+## Motivations for a new implementation
+
+ROS2 Dashing release is the first ROS2 release featuring a working intra-process communication mechanism.
+
+It is based on the creation of a ring buffer for each publisher and on the publication of meta-messages through
+the middleware layer. When a `Publisher` has to publish in-process, it will pass the message to the
+`IntraProcessManager`. Here the message will be stored in the ring buffer associated with the
+`Publisher`. In order to retrieve a message, the `IntraProcessManager` needs two pieces of information:
+the id of the publisher (in order to select the correct ring buffer) and the position of the message within
+the ring buffer. A meta-message with this information is created and sent through the ROS2 middleware to all
+the `Subscription`s, which can then retrieve the original message from the `IntraProcessManager`.
+
+![Current IPC Block Diagram](old_ipc.png)
+
+Several shortcomings of the current implementation are listed below.
+
+### Incomplete Quality of Service support
+
+The current implementation can't be used when the QoS durability value is set to `Transient Local`.
+
+The current implementation does not enforce the depth of the QoS history in a correct way. The reason is that
+there is a single ring buffer per `Publisher` and its size is equal to the depth of the `Publisher`'s history.
+A `Publisher` stores a message in the ring buffer and then it sends a meta-message to allow a `Subscription` to retrieve it.
+The `Subscription` correctly stores meta-messages up to the number indicated by its depth of the history, but,
+depending on the frequency at which messages are published and callbacks are triggered, it may happen that a
+meta-message processed from the `Subscription` does not correspond anymore to a valid message in the ring
+buffer, because it has been already overwritten.
+
+More details [here](https://index.ros.org/doc/ros2/Concepts/About-Quality-of-Service-Settings/).
+
+**TODO:** take into account also new QoS: Deadline, Liveliness and Lifespan
+[reference](https://github.com/ros2/design/pull/212).
+
+### Dependent on the RMW
+
+The current implementation has to send meta-messages from the `Publisher` to the `Subscription`s. This
+is done using the `rmw_publish` function, the implementation of which depends on the chosen middleware. This
+results in having the performance of a ROS2 application with IPC enabled to still heavily depend on the chosen
+RMW implementation.
+
+At the moment, all the ROS2 supported middlewares (Fast-RTPS, OpenSplice, Connext, DPS) do not make
+distinctions whether the `Publisher` and the `Subscription` are in the same process or not. In consequence,
```

Comment:
Looking at Fast-RTPS and DPS (and their RMW) I'm pretty sure that's correct.
It's also supported from experimental evaluations where having a publisher and a subscription in the same process or in two processes of the same machine does not affect the performances.

For what concerns OpenSplice, that statement is not correct, as this middleware uses the same ports for all the entities in the same process. This results in an almost instantaneous discovery, but I haven't seen any changes in latency for example.

No idea about Connext.

Reply:
Either way, they could do that if they wanted. I think it's worth separating whether or not they could make optimizations versus whether or not they do. Not being able to make these optimizations is a strong case for improving our intra-process system, but in the current situation, a valid question is "why not improve the rmw implementations using the information they have first"? I'm not saying we should do that, it's just a weak justification for adding more stuff into our intra-process pipeline instead, whereas if it were not possible for them to do those optimizations, that would be a strong reason to do it in rclcpp.

For example, a strong reason to do it in rclcpp (in my opinion) is that when we add type masquerading (using OpenCV Image or PCL point cloud instead of the ROS message type) it will be easier to support intra-process with those types if it is in rclcpp, though that doesn't imply that we use an envelope message or have our own queueing system.

Reply:
"Optimization" is not a binary operation though. There's no doubt that DDS vendors could do some optimizations and improve the efficiency of intra-process communication but there's only so much that can be done at the rmw layer. Unless some APIs are re-designed, there would always be the overhead of some additional copies. This overhead can be avoided entirely by doing this optimization in rclcpp directly without depending on the DDS vendor. IMO that is a strong enough reason since it guarantees a minimum level of performance irrespective of the DDS library.

Reply:
I agree, but my point is that it's not a strong supporting reason to put more complexity into rclcpp rather than try to improve the existing design. Since we haven't even tried to improve performance within the rmw layer. If performance (latency and cpu/memory usage) were improved for small messages within the same process at the rmw layer, then the drawbacks of using a "meta-message" decrease significantly, even if they will perhaps never be as good as a queue in rclcpp itself.

It's a bit of a moot point since the work is essentially already done, but adding things to the rclcpp layer adds complexity to a part of the system that is not optional (or at least not without removing code manually -- they don't have to use intra process after all).

I'm thinking towards certification efforts where they might have a much simpler or already certified rmw/middleware layer below.

To be clear, I think the approach described here is the best option, I'm just trying to make sure it's properly motivated as it does cost something now and in the future (in terms of maintenance). For example, each time we expose a QoS in ROS 2 we'll have to update the intra process system, and I think the proposed system will be more fragile/complex to update than the meta-message approach we have now.

Reply:
I agree that the proposed system will require an effort every time a new QoS is added.
However, I don't think that the current one handles this in a better way.

The RMW delivers the meta-messages enforcing the QoS, however if the IntraProcessManager "thinks" that a message has to delivered to N subscription, it will wait for all of them before releasing the memory in the ring buffer. This could be a potential source of problems.

Moreover, even if this proposal adds complexity to the rclcpp layer, I think that it's important that the intra-process communication is implemented in a way that is agnostic to the RMW.
Implementing it in the RMW will potentially lead to have different versions of intra-process communication for each RMW implementation...